File: python/0802.py
# Link: https://leetcode.com/problems/find-eventual-safe-states/description/
# Runtime: 832 ms
# Memory: 25.1 MB

import queue


class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        outdegree = [0] * len(graph)
        is_safe_node = [False] * len(graph)
        nodes_for_deleting = queue.Queue()
        connection_with_nodes = defaultdict(lambda: [])
        for i, connected_nodes in enumerate(graph):
            outdegree[i] = len(connected_nodes)
            if outdegree[i] == 0:
                is_safe_node[i] = True
                nodes_for_deleting.put(i)
            
            for node_index in connected_nodes:
                connection_with_nodes[node_index].append(i)
        
        while not nodes_for_deleting.empty():
            node_index = nodes_for_deleting.get()

            for node in connection_with_nodes[node_index]:
                outdegree[node] -= 1
                if outdegree[node] == 0:
                    nodes_for_deleting.put(node)
                    is_safe_node[node] = True
        
        safe_nodes = []
        for i, can_add in enumerate(is_safe_node):
            if can_add:
                safe_nodes.append(i)
        
        return safe_nodes

File: c++/0802.cpp
// Link: https://leetcode.com/problems/find-eventual-safe-states/description/
// Runtime: 270 ms
// Memory: 68 MB

class Solution {
public:
  std::vector<int>
  eventualSafeNodes(const std::vector<std::vector<int>> &graph) {
    std::vector<int> outdegree(graph.size());
    std::vector<bool> isSafeNode(graph.size());
    std::queue<int> nodesForDeleting;
    std::unordered_map<int, std::vector<int>> connectionWithNodes;
    for (int i = 0; i < static_cast<int>(graph.size()); i++) {
      outdegree[i] = static_cast<int>(graph[i].size());
      if (outdegree[i] == 0) {
        isSafeNode[i] = true;
        nodesForDeleting.push(i);
      }
      for (int j = 0; j < static_cast<int>(graph[i].size()); j++)
        connectionWithNodes[graph[i][j]].push_back(i);
    }

    while (!nodesForDeleting.empty()) {
      int nodeIndex = nodesForDeleting.front();
      nodesForDeleting.pop();

      for (const auto &value : connectionWithNodes[nodeIndex]) {
        outdegree[value]--;
        if (outdegree[value] == 0) {
          nodesForDeleting.push(value);
          isSafeNode[value] = true;
        }
      }
    }

    std::vector<int> safeNodes;
    for (int i = 0; i < static_cast<int>(isSafeNode.size()); i++)
      if (isSafeNode[i])
        safeNodes.push_back(i);

    return safeNodes;
  }
};
